### ISSUE TITLE
pick up `http_proxy` and `https_proxy` environment variables

### ISSUE THREAD

**Issue by toymil** [rustwasm/wasm-pack#1347]
## 💡 Feature description

Pick up `http_proxy` and `https_proxy` environment variables for all the download tasks (like downloading _chromedriver_ when executing `wasm-pack test --chrome --headless` for the first time).

This will prevent frustrating network problems like #605 and #1345 , since there is currently no way to set proxy for wasm-pack.

## Solution

From what I can tell by a quick dig in the source code, wasm-pack is using the `ureq` crate for all the download tasks?  If so, the solution should be straight forward, just upgrade the `ureq` dependency to at least version `2.8.0` and it should automatically pick up proxy related environment variables.

Related `ureq` issue and pr:
- https://github.com/algesten/ureq/issues/322
- https://github.com/algesten/ureq/pull/649

**Comment by toymil** [comment]
I just modified `Cargo.toml` from current master branch, upgraded `ureq` to `v2.9.0`, compiled and installed locally.  And now wasm-pack does pick up my `http(s)_proxy` environment variables, and successfully downloaded _chromedriver_.  Though I don't know if it break anything else.

**Comment by algesten** [comment]
I believe this is done. It's released in ureq 2.8.0 (and we're on 2.9.0 now).

**Comment by toymil** [comment]
I did the following:

- clone `ureq`, checkout `2.9.0`
  - modified to debug print proxy settings each time a request is made, to confirm if proxy envvars have been picked up or not.
- clone `binary-install`, checkout `v0.2.0`
  - change `Cargo.toml` to use local modified `ureq`
- clone `wasm-pack`, checkout `v0.12.1`
  - change `Cargo.toml` to use local modified `binary-install` and `ureq`
- install this local version of `wasm-pack` for tests

and the tests are negative, none of the network requests picked up proxy envvars...

Will do a bit more digging to see if I missed something.

**Comment by toymil** [comment]
I see that wasm-pack and binary-install all calls `ureq::get(..)` to download, which initializes an agent with the default setting `AgentBuilder::new().build()`, and by default `try_proxy_from_env` of `AgentBuilder` is set to `false`, so it ended up not using the proxy environment variables.

https://github.com/algesten/ureq/blob/260213b3e2a55e40f7132374034520455d497789/src/agent.rs#L252-L275

https://github.com/algesten/ureq/blob/260213b3e2a55e40f7132374034520455d497789/src/agent.rs#L323-L332

**Comment by toymil** [comment]
Changing `try_proxy_from_env` default to `true` does not solve the problem, **can confirm `Proxy::try_from_system()` did run**, but seems like `ureq` did not successfully pick up the environment variables @algesten .

I will continue this discussion on the ureq repo for now.

**Comment by algesten** [comment]
Yes. To use this, create an `Agent`, enable the environment option, and make the request from the agent.

**Comment by toymil** [comment]
Change `try_proxy_from_env` default to `true` made everything work.  Here is what I modified:

- clone `ureq`, checkout `2.9.0`
  - modified to debug print proxy settings each time a request is made, to confirm if proxy env vars have been picked up or not.
  - modified `AgentBuilder::new()` to enable `try_proxy_from_env` by default
- clone `binary-install`, checkout `v0.2.0`
  - change `Cargo.toml` to use local modified `ureq`
- clone `wasm-pack`, checkout `v0.12.1`
  - change `Cargo.toml` to use local modified `binary-install` and `ureq`
- install this local version of `wasm-pack` for tests

And wasm-pack successfully picked up my proxy env vars.

---

The most simple scenario would be for `ureq` to default `try_proxy_from_env` to `true`, then the changes needed here are:
- `wasm-pack` and `binary-install` both update the `ureq` dependency.
- optionally enable `socks-proxy` feature flag for `ureq`, so socks proxies in `http(s)_proxy` env var are also supported.

If not, then we need to manually construct `Agent` and call `AgentBuilder::try_proxy_from_env(mut self, do_try: bool)` to enable the behavior each time we need to download.  Since there is multiple places in the code where ureq are called, this would be tricky.

**Comment by toymil** [comment]
@algesten may I ask why `try_proxy_from_env` is default to `false`?  The comment wrote:

> The default is `false`, i.e. not detecting proxy from env since this is a potential security risk.

But most of the tooling I use does do this by default.

**Comment by toymil** [comment]
> But most of the tooling I use does do this by default. So why not enable it by default and provide an option for library users to disable instead?

I understand that it might always been the case for network libs to disable this behavior by default, and the reason why "most of the tooling I use does do this by default" is because the tooling devs manually enabled it.

Is this the case?

**Comment by algesten** [comment]
The motivation is hidden in the PR. I should probably bring it out in the docs. https://github.com/algesten/ureq/pull/649/files#r1290992185

**Comment by yaoshiu** [comment]
Hey, I just started using `wasm-pack`, and I’m not sure why this issue has been left unresolved for so long. From what I can see, it’s just one step away from being fixed.

The upstream `binary-install` crate has already merged the PR that adds support for the `http_proxy` environment variable. However, it’s been 7 months and there still hasn’t been a new release published.

I tried modifying the current `Cargo.toml` like this and compiled it locally. It works great, and `wasm-pack` now respects my `http_proxy` settings as expected:

```diff
diff --git a/Cargo.toml b/Cargo.toml
index ed36f30..ad0c84d 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -12,7 +12,7 @@ documentation = "https://rustwasm.github.io/wasm-pack/"

 [dependencies]
 anyhow = "1.0.68"
-binary-install = "0.4.1"
+binary-install = { git = "https://github.com/rustwasm/binary-install", version = "0.4.1" }
 cargo_metadata = "0.15.2"
 chrono = "0.4.23"
 console = "0.15.5"
```

I hope this helps the maintainers move things forward more quickly. Being unable to use a proxy is quite painful for developers like me who work in regions with poor network connectivity.

Thanks for all your hard work on this project!

**Comment by algesten** [comment]
ureq3 picks up the http_proxy and https_proxy variables. In a default `Agent` config, we do [`try_from_env()`](https://docs.rs/ureq/3.0.11/src/ureq/config.rs.html#881). That in turn checks [these variables](https://docs.rs/ureq/3.0.11/ureq/struct.Proxy.html#method.try_from_env).

If wasm-pack updates from 2 to 3, this would be solved.

**Comment by toymil** [comment]
You are right.  I didn't know about #1438 .

Although I don't like the idea of manually specify `try_proxy_from_env` at every call site, it does fix the problem.
